aeneas/src/x86-64/X86_64Darwin.v3

```diff
@@ -0,0 +1,273 @@
+		getRuntime(prog).addRoots();
+	}
+	private def getRuntime(prog: Program) -> MachRuntime {
+		return MachProgram.!(prog.tprog).runtime;
+	}
+	def emitDBG(compiler: Compiler, prog: Program) {
+		var rt = getRuntime(prog), mach = rt.mach;
+		mach.build(compiler, MachLoweringConfig.new());
+
+		var header = MachO_Header.new(true);
+		var pageAlign = rt.codeRegion.space.pageAlign;
+		header.cputype = MachO.CPU_TYPE_X86_64;
+		header.cpusubtype = 3;
+		header.filetype = MachO.MH_EXECUTE;
+
+		// protect page zero for trapping null accesses
+		//var pz = newSegmentLoad(header, "__NULL", 0);
+		//pz.vmsize = pageAlign.size;
+		//var cs = newSegmentLoad(header, "__TEXT", MachO.VM_PROT_RX);
+		//var ds = newSegmentLoad(header, "__DATA", MachO.VM_PROT_RW);
+		var w = MachDataWriter.new(rt.codeRegion.space.pageAlign,
+			0, 300 + mach.numMethods * 50);
+		// skip mach-o headers
+		w.skipN(header.totalsize);
+		// skip page zero
+		w.skipPage();
+		// encode header at position 0 now that all addresses and offsets are known
+		header.encode(w.at(0));
+		header.encodeCmds(w);
+		// open the output file
+		var file = compiler.getOutputFileName(prog.name(), null);
+		var fd = System.fileOpen(file, false);
+		if (fd < 0) return prog.ERROR.OutputError(file);
+		// write the entire file from the buffer array
+		System.fileWriteK(fd, w.data, 0, w.end());
+		System.fileClose(fd);
+		// change permissions to make binary executable
+		compiler.makeExecutable(file);
+	}
+	def emit(compiler: Compiler, prog: Program) {
+		var rt = getRuntime(prog), mach = rt.mach;
+		mach.build(compiler, MachLoweringConfig.new());
+
+		var header = MachO_Header.new(true);
+		var pageAlign = rt.codeRegion.space.pageAlign;
+		header.cputype = MachO.CPU_TYPE_X86_64;
+		header.cpusubtype = 3;
+		header.filetype = MachO.MH_EXECUTE;
+
+		// protect page zero for trapping null accesses
+		var pz = newSegmentLoad(header, "__NULL", 0);
+		pz.vmsize = pageAlign.size;
+		// code (text) segment starts at page 1
+		var cs = newSegmentLoad(header, "__TEXT", MachO.VM_PROT_RX);
+		// data segment immediately follows code segment
+		var ds = newSegmentLoad(header, "__DATA", MachO.VM_PROT_RW);
+		// initial thread state for registers (includes starting IP)
+		var ts = MachO_X86_64ThreadState.new();
+		header.addCmd(ts);
+		// ex segment for trapping explicit checks
+		var ex = if(rt.src != null, newSegmentLoad(header, "__EX", 0));
+		// create the buffer
+		var w = MachDataWriter.new(rt.codeRegion.space.pageAlign,
+			0, 300 + mach.numMethods * 50);
+		// skip mach-o headers
+		w.skipN(header.totalsize);
+		// skip page zero
+		w.skipPage();
+		// generate code and runtime tables into buffer
+		// encodeCode(w, cs, ts);
+		var backend = X86_64DarwinBackend.new(compiler, prog, mach, w, test);
+		ts.eip = w.endAddr();
+		cs.vmaddr = w.endPageAddr();
+		rt.recordCodeStart(ts.eip);
+		backend.genAllCode();
+		var size = w.end();
+		if (size < pageAlign.size) {
```

Comment:
I think you can remove this `if` for now; we'll add this back if it is an issue.